src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.ValueTaskSource.Windows.cs

```diff
@@ -1,22 +1,49 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System;
 using System.Buffers;
 using System.Diagnostics;
+using System.IO;
 using System.Threading;
 using System.Threading.Tasks.Sources;
 
-namespace System.IO.Strategies
+namespace Microsoft.Win32.SafeHandles
 {
-    internal sealed partial class AsyncWindowsFileStreamStrategy : WindowsFileStreamStrategy
+    public sealed partial class SafeFileHandle : SafeHandleZeroOrMinusOneIsInvalid
     {
+        private ValueTaskSource? _reusableValueTaskSource; // reusable ValueTaskSource that is currently NOT being used
+
+        // Rent the reusable ValueTaskSource, or create a new one to use if we couldn't get one (which
+        // should only happen on first use or if the FileStream is being used concurrently).
+        internal ValueTaskSource GetValueTaskSource() => Interlocked.Exchange(ref _reusableValueTaskSource, null) ?? new ValueTaskSource(this);
```

Comment:
To be honest with you I've added it because I was afraid of your pushback based on the following comment: https://github.com/dotnet/runtime/issues/24847#issuecomment-808367922

> I disagree it will significantly increase the complexity (especially if the "pool" is a single field storing one instance to reuse), and I disagree parallel usage will be the most common use case (I agree it will be a use case).

Would it be acceptable to keep it as it is and change it after we receive some feedback from the users?

Or I can modify the code to do the following:

* FileStream - keep caching
* Random access for single buffer - don't cache
* Random access for multiple buffers - create a single instance and reuse it for all the buffers

@stephentoub what do you think?

Reply:
> Would it be acceptable to keep it as it is and change it after we receive some feedback from the users?

It depends on the expected use case.  You can measure to confirm or refute my hypothesis, but I'd expect this to be very expensive for anything but the first concurrent operation.  If you expect concurrent usage to be a meaningful use case, and if I'm right that it's fairly expensive as a throwaway instance, then you might find it beneficial to use a different implementation for the accesses the are unable to use the singleton.

Reply:
> You can measure to confirm or refute my hypothesis

Currently benchmarking new APIs is a PITA. I would prefer to merge it as it is, add benchmarks for the new APIs as soon as they are available in the SDK, and then send a separate PR with improvements. @stephentoub is that acceptable?

FWIW when I was implementing these APIs and writing the tests I got a feeling that users are most probably going to use:
* overloads that accept a single buffer in a non-concurrent way
* overloads that accept multiple buffers in a concurrent way

Using the scatter/gather APIs is just much more convenient when you have multiple buffers.

Reply:
> @stephentoub is that acceptable?

Yup

> overloads that accept multiple buffers in a concurrent way

Can you elaborate on what led to that conclusion?

> Using the scatter/gather APIs is just much more convenient when you have multiple buffers.

Maybe; it depends on the data structure you have storing those buffers. e.g. I'd be interested in seeing a sample that uses one of these with a ReadOnlySequence<byte>.  I'm also not clear on why having multiple buffers makes it more likely you'd want to issue concurrent calls via RandomAccess on the same file handle.

Reply:
> Can you elaborate on what led to that conclusion?

It simply hard for me to come up with a scenario where I would like to perform multiple reads|writes that are not consecutive. 

Example: I have 3 buffers with some content and I want to store them to the disk: A, B, C.

If I start 3 concurrent write operations:
```ini
(buffer: A, offset: 0)
(buffer: B, offset: A.Length)
(buffer: C, offset: A.Length+B.Length)
```

I need to: 
* hardcode the code to perform 3 awaits or allocate an array of tasks and cast each VT to Task to be able to call `Task.WaitAll` or similar API
* each call can end up with partial success that requires special handling
* do the math in the right way to calculate the indexes

With Scatter/Gather all I have is a single await and single check for partial success. I might also get better perf (single sys call).

Reply:
I think we're talking past each other.  A single call to Read(multipleBuffers) is not concurrent usage of RandomAccess.

Reply:
I've created an issue to keep track of it: https://github.com/dotnet/performance/issues/1840